src/transformers/hf_argparser.py

```diff
@@ -70,93 +70,100 @@ def __init__(self, dataclass_types: Union[DataClassType, Iterable[DataClassType]
         for dtype in self.dataclass_types:
             self._add_dataclass_arguments(dtype)
 
+    @staticmethod
+    def _parse_dataclass_field(parser: ArgumentParser, field: dataclasses.Field):
+        field_name = f"--{field.name}"
+        kwargs = field.metadata.copy()
+        # field.metadata is not used at all by Data Classes,
+        # it is provided as a third-party extension mechanism.
+        if isinstance(field.type, str):
+            raise RuntimeError(
+                "Unresolved type detected, which should have been done with the help of "
+                "`typing.get_type_hints` method by default"
+            )
+
+        origin_type = getattr(field.type, "__origin__", field.type)
+        if origin_type is Union:
+            if len(field.type.__args__) != 2 or type(None) not in field.type.__args__:
+                raise ValueError("Only `Union[X, NoneType]` (i.e., `Optional[X]`) is allowed for `Union`")
+            if bool not in field.type.__args__:
+                # filter `NoneType` in Union (except for `Union[bool, NoneType]`)
+                field.type = next(
+                    filter(lambda t: not isinstance(None, getattr(t, "__origin__", t)), field.type.__args__)
+                )
```

Comment:
This line is way to hard to follow. At this stage we know we have only two types, one being NoneType so I think we just want to do
```suggestion
                field.type = field.type.__args__[0] if field.type.__args__[0] != type(None) else field.type.__args__[1]
```